app/src/main/java/com/example/habitsmasher/ui/history/AddHabitEventDialog.java

```diff
@@ -0,0 +1,237 @@
+                    HabitEvent newEvent = new HabitEvent(_eventDate, _eventCommentInput, _selectedImage);
+                    _habitEventListFragment.addNewHabitEvent(newEvent);
+                    _habitEventListFragment.addHabitEventToDatabase(newEvent.getStartDate(), newEvent.getComment(), newEvent.getId(), newEvent.getPictureUri(), _username);
+                    getDialog().dismiss();
+                }
+
+            }
+        });
+
+        // Add listener to image view
+        _eventPictureView.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                // Open gallery to let user pick photo
+                Intent pickPhoto = new Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
+                startActivityForResult(pickPhoto, 1);
+            }
+        });
+
+        return view;
+    }
+
+
+    /**
+     * Reference: https://stackoverflow.com/questions/10165302/dialog-to-pick-image-from-gallery-or-from-camera
+     * Override onActivityResult to handle when user has selected image
+     * @param requestCode
+     * @param resultCode
+     * @param imageReturnedIntent
+     */
+    @Override
+    public void onActivityResult(int requestCode, int resultCode, Intent imageReturnedIntent) {
+        super.onActivityResult(requestCode, resultCode, imageReturnedIntent);
+        switch(requestCode) {
+            case 0:
+            case 1:
+                if (resultCode == RESULT_OK) {
+                    // Set selected picture
+                    _selectedImage = imageReturnedIntent.getData();
+                    _eventPictureView.setImageURI(_selectedImage);
+                }
+                break;
+        }
+
+    }
+
+    /**
+     * Opens date picker
+     */
+    private void openDatePickerDialog(){
+        DatePickerDialog datePickerDialog = new DatePickerDialog(
+                getActivity(),
+                this,
+                Calendar.getInstance().get(Calendar.YEAR),
+                Calendar.getInstance().get(Calendar.MONTH),
+                Calendar.getInstance().get(Calendar.DAY_OF_MONTH)
+        );
+        datePickerDialog.show();
+    }
+
+    /**
+     * Handles when a date is selected
+     * @param view
+     * @param year
+     * @param month
+     * @param day
+     */
+    @Override
+    public void onDateSet(DatePicker view, int year, int month, int day) {
+        //1 is added to the month we get from the DatePickerDialog
+        // because DatePickerDialog returns values between 0 and 11,
+        // which is not really helpful for users.
+        int correctedMonth = month + 1;
+        String date = day + "/" + correctedMonth + "/" + year;
+        _eventDateText.setText(date);
+    }
+
+    /**
+     * Function to check if data of a potential new habit event is valid
+     * @return true or false depending on data validity
+     */
+    private boolean checkNewHabitEventIsValid() {
```

Comment:
I would prefer this to be **extracted** to its own class, `HabitEventValidator`, and see some specific unit tests for all input combinations. You can take "inspiration" from the `HabiValidator`, and also its associated tests.